### PR TITLE
Re-sync unassigned egress IPs for CloudPrivateIPConfig deletes

### PR DESCRIPTION
**Pull request by alexanderConstantinescu**
/cc @kyrtapz 

This turned out to be more complicated than expected, so I started looking at the code, fiddling with it and by the end writing a solution. 

Problem: on public clouds the amount of private IP addresses that can be assigned to nodes are limited. Hence there's a chance some `EgressIP` objects might not get all of their egress IPs assigned. Two simple cases demonstrating this is:

1. Assume we have 4 nodes with capacity 1 and two `EgressIP` objects with spec EIP1=[1,2,3,4] and EIP2=[5]. So the status for each could be EIP1=[1,2,3,4] and EIP2=[]. Now say the user changes the spec on EIP1 to [1,2,3] and we remove 4 from one node. When the `CloudPrivateIPConfig` associated with 4 is deleted, we want to directly assign 5 from EIP2 to the node. 
2. Assume we have 1 node with capacity 4 and 5 `EgressIP` objects with spec  EIP1=[2],  EIP1=[2] , EIP1=[3] , EIP1=[4], EIP1=[5]. So the status for each could be:  EIP1=[2],  EIP1=[2] , EIP1=[3] , EIP1=[4], EIP1=[]. Now say a user deleted EIP4: when the `CloudPrivateIPConfig` associated with 4 is deleted, we want to directly assign EIP5 to the node.
3. For both 1. and 2. above we also need to take into account the case where: when the `CloudPrivateIPConfig` associated with 4 is deleted, we need to check that the `CloudPrivateIPConfig` associated with EIP2/EIP5 is not waiting on assignment somewhere else (in the pending queue). If that is the case we might assign EIP2/EIP5 to two different nodes simultaneously. 

This PR thus introduces a cache tracking all pending assignments, and only resyncs `EgressIPs` which have egress IPs which are not pending. 

This PR replaces: #2774 

**- What this PR does and why is it needed**

**- Special notes for reviewers**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: go-controller/pkg/ovn/egressip.go [hunk 1]
```diff
@@ -605,10 +615,95 @@ func (oc *Controller) reconcileCloudPrivateIPConfig(old, new *ocpcloudnetworkapi
...
+			// all existing objects, for which no CR was issued.
+			klog.V(5).Infof("No pending operation found for EgressIP: %s while processing created CloudPrivateIPConfig", egressIPName)
+			return nil
+		}
+		op, exists := ops[egressIPString]
+		if !exists {
+			klog.V(5).Infof("Pending operations found for EgressIP: %s, but not for the created CloudPrivateIPConfig: %s", egressIPName, egressIPString)
+			return nil
+		}
+		if (op.toAdd != "" && op.toDelete != "") || (op.toAdd != "" && op.toDelete == "") {
```
Comment (kyrtapz) [inline go-controller/pkg/ovn/egressip.go:639]: isn't it the same as `if op.toAdd != ""`?

Comment (alexanderConstantinescu) [reply go-controller/pkg/ovn/egressip.go:639]: It is. I will update it

**Review by flavio-fernandes — COMMENTED** [review]
/lgtm

**Pull request merged**